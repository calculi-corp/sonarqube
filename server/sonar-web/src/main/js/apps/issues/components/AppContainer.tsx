/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import withBranchStatusActions from '../../../app/components/branch-status/withBranchStatusActions';
import withCurrentUserContext from '../../../app/components/current-user/withCurrentUserContext';
import { PageContext } from '../../../app/components/indexation/PageUnavailableDueToIndexation';
import withIndexationGuard from '../../../components/hoc/withIndexationGuard';
import { withRouter } from '../../../components/hoc/withRouter';
import { lazyLoadComponent } from '../../../components/lazyLoadComponent';

const IssuesAppContainer = lazyLoadComponent(() => import('./IssuesApp'), 'IssuesAppContainer');

export default withIndexationGuard(
  withRouter(withCurrentUserContext(withBranchStatusActions(IssuesAppContainer))),
  PageContext.Issues
);
