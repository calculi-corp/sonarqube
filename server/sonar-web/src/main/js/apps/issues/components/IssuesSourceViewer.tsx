/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import { scrollToElement } from '../../../helpers/scrolling';
import { BranchLike } from '../../../types/branch-like';
import { Issue } from '../../../types/types';
import CrossComponentSourceViewer from '../crossComponentSourceViewer/CrossComponentSourceViewer';
import { getLocations, getSelectedLocation } from '../utils';

interface Props {
  branchLike: BranchLike | undefined;
  issues: Issue[];
  locationsNavigator: boolean;
  onIssueChange: (issue: Issue) => void;
  onIssueSelect: (issueKey: string) => void;
  onLocationSelect: (index: number) => void;
  openIssue: Issue;
  selectedFlowIndex: number | undefined;
  selectedLocationIndex: number | undefined;
}

export default class IssuesSourceViewer extends React.PureComponent<Props> {
  node?: HTMLElement | null;

  componentDidUpdate(prevProps: Props) {
    const { openIssue, selectedLocationIndex } = this.props;

    // Scroll back to the issue when the selected location is set to -1
    const shouldScrollBackToIssue =
      selectedLocationIndex === -1 && selectedLocationIndex !== prevProps.selectedLocationIndex;
    if (
      prevProps.openIssue.component === openIssue.component &&
      (prevProps.openIssue !== openIssue || shouldScrollBackToIssue)
    ) {
      this.scrollToIssue();
    }
  }

  scrollToIssue = (smooth = true) => {
    if (this.node) {
      const element = this.node.querySelector(`[data-issue="${this.props.openIssue.key}"]`);
      if (element) {
        this.handleScroll(element, undefined, smooth);
      }
    }
  };

  handleScroll = (element: Element, offset = window.innerHeight / 2, smooth = true) => {
    scrollToElement(element, { topOffset: offset - 100, bottomOffset: offset, smooth });
  };

  handleLoaded = () => {
    this.scrollToIssue(false);
  };

  render() {
    const { openIssue, selectedFlowIndex, selectedLocationIndex } = this.props;

    const locations = getLocations(openIssue, selectedFlowIndex).map((loc, index) => {
      loc.index = index;
      return loc;
    });
    const selectedLocation = getSelectedLocation(
      openIssue,
      selectedFlowIndex,
      selectedLocationIndex
    );

    const highlightedLocationMessage =
      this.props.locationsNavigator && selectedLocationIndex !== undefined
        ? selectedLocation && { index: selectedLocationIndex, text: selectedLocation.msg }
        : undefined;

    return (
      <div ref={node => (this.node = node)}>
        <CrossComponentSourceViewer
          branchLike={this.props.branchLike}
          highlightedLocationMessage={highlightedLocationMessage}
          issue={openIssue}
          issues={this.props.issues}
          locations={locations}
          onIssueChange={this.props.onIssueChange}
          onIssueSelect={this.props.onIssueSelect}
          onLoaded={this.handleLoaded}
          onLocationSelect={this.props.onLocationSelect}
          scroll={this.handleScroll}
          selectedFlowIndex={selectedFlowIndex}
        />
      </div>
    );
  }
}
