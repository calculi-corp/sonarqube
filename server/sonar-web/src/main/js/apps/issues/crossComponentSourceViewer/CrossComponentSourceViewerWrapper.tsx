/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { findLastIndex, keyBy } from 'lodash';
import * as React from 'react';
import { getComponentForSourceViewer, getDuplications, getSources } from '../../../api/components';
import { getIssueFlowSnippets } from '../../../api/issues';
import DuplicationPopup from '../../../components/SourceViewer/components/DuplicationPopup';
import {
  filterDuplicationBlocksByLine,
  getDuplicationBlocksForIndex,
  isDuplicationBlockInRemovedComponent
} from '../../../components/SourceViewer/helpers/duplications';
import {
  duplicationsByLine as getDuplicationsByLine,
  issuesByComponentAndLine
} from '../../../components/SourceViewer/helpers/indexing';
import { SourceViewerContext } from '../../../components/SourceViewer/SourceViewerContext';
import { Alert } from '../../../components/ui/Alert';
import DeferredSpinner from '../../../components/ui/DeferredSpinner';
import { WorkspaceContext } from '../../../components/workspace/context';
import { getBranchLikeQuery } from '../../../helpers/branch-like';
import { throwGlobalError } from '../../../helpers/error';
import { translate } from '../../../helpers/l10n';
import { HttpStatus } from '../../../helpers/request';
import { BranchLike } from '../../../types/branch-like';
import { isFile } from '../../../types/component';
import {
  Dict,
  DuplicatedFile,
  Duplication,
  FlowLocation,
  Issue,
  SnippetsByComponent,
  SourceViewerFile
} from '../../../types/types';
import ComponentSourceSnippetGroupViewer from './ComponentSourceSnippetGroupViewer';
import { getPrimaryLocation, groupLocationsByComponent } from './utils';

interface Props {
  branchLike: BranchLike | undefined;
  highlightedLocationMessage?: { index: number; text: string | undefined };
  issue: Issue;
  issues: Issue[];
  locations: FlowLocation[];
  onIssueChange: (issue: Issue) => void;
  onIssueSelect: (issueKey: string) => void;
  onLoaded?: () => void;
  onLocationSelect: (index: number) => void;
  scroll?: (element: HTMLElement) => void;
  selectedFlowIndex: number | undefined;
}

interface State {
  components: Dict<SnippetsByComponent>;
  duplicatedFiles?: Dict<DuplicatedFile>;
  duplications?: Duplication[];
  duplicationsByLine: { [line: number]: number[] };
  issuePopup?: { issue: string; name: string };
  loading: boolean;
  notAccessible: boolean;
}

export default class CrossComponentSourceViewerWrapper extends React.PureComponent<Props, State> {
  mounted = false;
  state: State = {
    components: {},
    duplicationsByLine: {},
    loading: true,
    notAccessible: false
  };

  componentDidMount() {
    this.mounted = true;
    this.fetchIssueFlowSnippets();
  }

  componentDidUpdate(prevProps: Props) {
    if (prevProps.issue.key !== this.props.issue.key) {
      this.fetchIssueFlowSnippets();
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  fetchDuplications = (component: string) => {
    getDuplications({
      key: component,
      ...getBranchLikeQuery(this.props.branchLike)
    }).then(
      r => {
        if (this.mounted) {
          this.setState({
            duplicatedFiles: r.files,
            duplications: r.duplications,
            duplicationsByLine: getDuplicationsByLine(r.duplications)
          });
        }
      },
      () => {}
    );
  };

  async fetchIssueFlowSnippets() {
    const { issue, branchLike } = this.props;
    this.setState({ loading: true });

    try {
      const components = await getIssueFlowSnippets(issue.key);
      if (components[issue.component] === undefined) {
        const issueComponent = await getComponentForSourceViewer({
          component: issue.component,
          ...getBranchLikeQuery(branchLike)
        });
        components[issue.component] = { component: issueComponent, sources: [] };
        if (isFile(issueComponent.q)) {
          const sources = await getSources({
            key: issueComponent.key,
            ...getBranchLikeQuery(branchLike),
            from: 1,
            to: 10
          }).then(lines => keyBy(lines, 'line'));
          components[issue.component].sources = sources;
        }
      }
      if (this.mounted) {
        this.setState({
          components,
          issuePopup: undefined,
          loading: false
        });
        if (this.props.onLoaded) {
          this.props.onLoaded();
        }
      }
    } catch (response) {
      const rsp = response as Response;
      if (rsp.status !== HttpStatus.Forbidden) {
        throwGlobalError(response);
      }
      if (this.mounted) {
        this.setState({ loading: false, notAccessible: rsp.status === HttpStatus.Forbidden });
      }
    }
  }

  handleIssuePopupToggle = (issue: string, popupName: string, open?: boolean) => {
    this.setState((state: State) => {
      const samePopup =
        state.issuePopup && state.issuePopup.name === popupName && state.issuePopup.issue === issue;
      if (open !== false && !samePopup) {
        return { issuePopup: { issue, name: popupName } };
      } else if (open !== true && samePopup) {
        return { issuePopup: undefined };
      }
      return null;
    });
  };

  renderDuplicationPopup = (component: SourceViewerFile, index: number, line: number) => {
    const { duplicatedFiles, duplications } = this.state;

    if (!component || !duplicatedFiles) {
      return null;
    }

    const blocks = getDuplicationBlocksForIndex(duplications, index);

    return (
      <WorkspaceContext.Consumer>
        {({ openComponent }) => (
          <DuplicationPopup
            blocks={filterDuplicationBlocksByLine(blocks, line)}
            branchLike={this.props.branchLike}
            duplicatedFiles={duplicatedFiles}
            inRemovedComponent={isDuplicationBlockInRemovedComponent(blocks)}
            openComponent={openComponent}
            sourceViewerFile={component}
          />
        )}
      </WorkspaceContext.Consumer>
    );
  };

  render() {
    const { loading, notAccessible } = this.state;

    if (loading) {
      return (
        <div>
          <DeferredSpinner />
        </div>
      );
    }

    if (notAccessible) {
      return (
        <Alert className="spacer-top" variant="warning">
          {translate('code_viewer.no_source_code_displayed_due_to_security')}
        </Alert>
      );
    }

    const { issue, locations } = this.props;
    const { components, duplications, duplicationsByLine } = this.state;
    const issuesByComponent = issuesByComponentAndLine(this.props.issues);
    const locationsByComponent = groupLocationsByComponent(issue, locations, components);

    const lastOccurenceOfPrimaryComponent = findLastIndex(
      locationsByComponent,
      ({ component }) => component.key === issue.component
    );

    if (components[issue.component] === undefined) {
      return null;
    }

    return (
      <div>
        {locationsByComponent.map((snippetGroup, i) => {
          return (
            <SourceViewerContext.Provider
              // eslint-disable-next-line react/no-array-index-key
              key={`${issue.key}-${this.props.selectedFlowIndex || 0}-${i}`}
              value={{ branchLike: this.props.branchLike, file: snippetGroup.component }}>
              <ComponentSourceSnippetGroupViewer
                branchLike={this.props.branchLike}
                duplications={duplications}
                duplicationsByLine={duplicationsByLine}
                highlightedLocationMessage={this.props.highlightedLocationMessage}
                issue={issue}
                issuePopup={this.state.issuePopup}
                issuesByLine={issuesByComponent[snippetGroup.component.key] || {}}
                isLastOccurenceOfPrimaryComponent={i === lastOccurenceOfPrimaryComponent}
                lastSnippetGroup={i === locationsByComponent.length - 1}
                loadDuplications={this.fetchDuplications}
                locations={snippetGroup.locations || []}
                onIssueChange={this.props.onIssueChange}
                onIssueSelect={this.props.onIssueSelect}
                onIssuePopupToggle={this.handleIssuePopupToggle}
                onLocationSelect={this.props.onLocationSelect}
                renderDuplicationPopup={this.renderDuplicationPopup}
                scroll={this.props.scroll}
                snippetGroup={snippetGroup}
              />
            </SourceViewerContext.Provider>
          );
        })}

        {locationsByComponent.length === 0 && (
          <ComponentSourceSnippetGroupViewer
            branchLike={this.props.branchLike}
            duplications={duplications}
            duplicationsByLine={duplicationsByLine}
            highlightedLocationMessage={this.props.highlightedLocationMessage}
            issue={issue}
            issuePopup={this.state.issuePopup}
            issuesByLine={issuesByComponent[issue.component] || {}}
            isLastOccurenceOfPrimaryComponent={true}
            lastSnippetGroup={true}
            loadDuplications={this.fetchDuplications}
            locations={[]}
            onIssueChange={this.props.onIssueChange}
            onIssueSelect={this.props.onIssueSelect}
            onIssuePopupToggle={this.handleIssuePopupToggle}
            onLocationSelect={this.props.onLocationSelect}
            renderDuplicationPopup={this.renderDuplicationPopup}
            scroll={this.props.scroll}
            snippetGroup={{
              locations: [getPrimaryLocation(issue)],
              ...components[issue.component]
            }}
          />
        )}
      </div>
    );
  }
}
