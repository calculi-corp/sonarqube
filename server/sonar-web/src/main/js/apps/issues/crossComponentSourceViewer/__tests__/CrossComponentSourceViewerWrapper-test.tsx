/*
 * SonarQube
 * Copyright (C) 2009-2022 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import { shallow } from 'enzyme';
import * as React from 'react';
import { getDuplications } from '../../../../api/components';
import { getIssueFlowSnippets } from '../../../../api/issues';
import {
  mockSnippetsByComponent,
  mockSourceLine,
  mockSourceViewerFile
} from '../../../../helpers/mocks/sources';
import { mockFlowLocation, mockIssue } from '../../../../helpers/testMocks';
import { waitAndUpdate } from '../../../../helpers/testUtils';
import CrossComponentSourceViewerWrapper from '../CrossComponentSourceViewerWrapper';

jest.mock('../../../../api/issues', () => {
  const { mockSnippetsByComponent } = jest.requireActual('../../../../helpers/mocks/sources');
  return {
    getIssueFlowSnippets: jest.fn().mockResolvedValue({ 'main.js': mockSnippetsByComponent() })
  };
});

jest.mock('../../../../api/components', () => ({
  getDuplications: jest.fn().mockResolvedValue({}),
  getComponentForSourceViewer: jest.fn().mockResolvedValue({})
}));

beforeEach(() => {
  jest.clearAllMocks();
});

it('should render correctly', async () => {
  let wrapper = shallowRender();
  expect(wrapper).toMatchSnapshot();

  await waitAndUpdate(wrapper);
  expect(wrapper).toMatchSnapshot();

  wrapper = shallowRender({ issue: mockIssue(true, { component: 'test.js', key: 'unknown' }) });
  await waitAndUpdate(wrapper);

  expect(wrapper).toMatchSnapshot('no component found');
});

it('Should fetch data', async () => {
  const wrapper = shallowRender();
  wrapper.instance().fetchIssueFlowSnippets();
  await waitAndUpdate(wrapper);
  expect(getIssueFlowSnippets).toHaveBeenCalledWith('1');
  expect(wrapper.state('components')).toEqual(
    expect.objectContaining({ 'main.js': mockSnippetsByComponent() })
  );

  (getIssueFlowSnippets as jest.Mock).mockClear();
  wrapper.setProps({ issue: mockIssue(true, { key: 'foo' }) });
  expect(getIssueFlowSnippets).toBeCalledWith('foo');
});

it('Should handle no access rights', async () => {
  (getIssueFlowSnippets as jest.Mock).mockRejectedValueOnce({ status: 403 });

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  expect(wrapper.state().notAccessible).toBe(true);
  expect(wrapper).toMatchSnapshot();
});

it('should handle issue popup', () => {
  const wrapper = shallowRender();
  // open
  wrapper.instance().handleIssuePopupToggle('1', 'popup1');
  expect(wrapper.state('issuePopup')).toEqual({ issue: '1', name: 'popup1' });

  // close
  wrapper.instance().handleIssuePopupToggle('1', 'popup1');
  expect(wrapper.state('issuePopup')).toBeUndefined();
});

it('should handle duplication popup', async () => {
  const files = { b: { key: 'b', name: 'B.tsx', project: 'foo', projectName: 'Foo' } };
  const duplications = [{ blocks: [{ _ref: '1', from: 1, size: 2 }] }];
  (getDuplications as jest.Mock).mockResolvedValueOnce({ duplications, files });

  const wrapper = shallowRender();
  await waitAndUpdate(wrapper);

  wrapper.find('ComponentSourceSnippetGroupViewer').prop<Function>('loadDuplications')(
    'foo',
    mockSourceLine()
  );

  await waitAndUpdate(wrapper);
  expect(getDuplications).toHaveBeenCalledWith({ key: 'foo' });
  expect(wrapper.state('duplicatedFiles')).toEqual(files);
  expect(wrapper.state('duplications')).toEqual(duplications);
  expect(wrapper.state('duplicationsByLine')).toEqual({ '1': [0], '2': [0] });

  expect(
    wrapper.find('ComponentSourceSnippetGroupViewer').prop<Function>('renderDuplicationPopup')(
      mockSourceViewerFile(),
      0,
      16
    )
  ).toMatchSnapshot();
});

function shallowRender(props: Partial<CrossComponentSourceViewerWrapper['props']> = {}) {
  return shallow<CrossComponentSourceViewerWrapper>(
    <CrossComponentSourceViewerWrapper
      branchLike={undefined}
      highlightedLocationMessage={undefined}
      issue={mockIssue(true, { key: '1' })}
      issues={[]}
      locations={[mockFlowLocation()]}
      onIssueChange={jest.fn()}
      onLoaded={jest.fn()}
      onIssueSelect={jest.fn()}
      onLocationSelect={jest.fn()}
      scroll={jest.fn()}
      selectedFlowIndex={0}
      {...props}
    />
  );
}
