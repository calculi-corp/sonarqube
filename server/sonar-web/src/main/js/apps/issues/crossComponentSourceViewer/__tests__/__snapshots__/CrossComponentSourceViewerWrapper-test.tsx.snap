// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should handle no access rights 1`] = `
<Alert
  className="spacer-top"
  variant="warning"
>
  code_viewer.no_source_code_displayed_due_to_security
</Alert>
`;

exports[`should handle duplication popup 1`] = `
<Context.Consumer>
  [Function]
</Context.Consumer>
`;

exports[`should render correctly 1`] = `
<div>
  <DeferredSpinner />
</div>
`;

exports[`should render correctly 2`] = `
<div>
  <ContextProvider
    key="1-0-0"
    value={
      Object {
        "branchLike": undefined,
        "file": Object {
          "canMarkAsFavorite": true,
          "fav": false,
          "key": "project:main.js",
          "longName": "main.js",
          "measures": Object {
            "coverage": "85.2",
            "duplicationDensity": "1.0",
            "issues": "12",
            "lines": "56",
          },
          "name": "main.js",
          "path": "main.js",
          "project": "project",
          "projectName": "MyProject",
          "q": "FIL",
          "uuid": "foo-bar",
        },
      }
    }
  >
    <ComponentSourceSnippetGroupViewer
      duplicationsByLine={Object {}}
      isLastOccurenceOfPrimaryComponent={false}
      issue={
        Object {
          "actions": Array [],
          "component": "main.js",
          "componentLongName": "main.js",
          "componentQualifier": "FIL",
          "componentUuid": "foo1234",
          "creationDate": "2017-03-01T09:36:01+0100",
          "flows": Array [
            Array [
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
            ],
            Array [
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
            ],
          ],
          "fromHotspot": false,
          "key": "1",
          "line": 25,
          "message": "Reduce the number of conditional operators (4) used in the expression",
          "project": "myproject",
          "projectKey": "foo",
          "projectName": "Foo",
          "rule": "javascript:S1067",
          "ruleName": "foo",
          "secondaryLocations": Array [
            Object {
              "component": "main.js",
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
            Object {
              "component": "main.js",
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
          ],
          "severity": "MAJOR",
          "status": "OPEN",
          "textRange": Object {
            "endLine": 26,
            "endOffset": 15,
            "startLine": 25,
            "startOffset": 0,
          },
          "transitions": Array [],
          "type": "BUG",
        }
      }
      issuesByLine={Object {}}
      lastSnippetGroup={true}
      loadDuplications={[Function]}
      locations={
        Array [
          Object {
            "component": "main.js",
            "index": 0,
            "textRange": Object {
              "endLine": 2,
              "endOffset": 2,
              "startLine": 1,
              "startOffset": 1,
            },
          },
        ]
      }
      onIssueChange={[MockFunction]}
      onIssuePopupToggle={[Function]}
      onIssueSelect={[MockFunction]}
      onLocationSelect={[MockFunction]}
      renderDuplicationPopup={[Function]}
      scroll={[MockFunction]}
      snippetGroup={
        Object {
          "component": Object {
            "canMarkAsFavorite": true,
            "fav": false,
            "key": "project:main.js",
            "longName": "main.js",
            "measures": Object {
              "coverage": "85.2",
              "duplicationDensity": "1.0",
              "issues": "12",
              "lines": "56",
            },
            "name": "main.js",
            "path": "main.js",
            "project": "project",
            "projectName": "MyProject",
            "q": "FIL",
            "uuid": "foo-bar",
          },
          "locations": Array [
            Object {
              "component": "main.js",
              "index": 0,
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
          ],
          "sources": Object {
            "16": Object {
              "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
              "coverageStatus": "covered",
              "coveredConditions": 2,
              "duplicated": false,
              "isNew": true,
              "line": 16,
              "scmAuthor": "simon.brandhof@sonarsource.com",
              "scmDate": "2018-12-11T10:48:39+0100",
              "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
            },
          },
        }
      }
    />
  </ContextProvider>
</div>
`;

exports[`should render correctly: no component found 1`] = `
<div>
  <ContextProvider
    key="unknown-0-0"
    value={
      Object {
        "branchLike": undefined,
        "file": Object {},
      }
    }
  >
    <ComponentSourceSnippetGroupViewer
      duplicationsByLine={Object {}}
      isLastOccurenceOfPrimaryComponent={false}
      issue={
        Object {
          "actions": Array [],
          "component": "test.js",
          "componentLongName": "main.js",
          "componentQualifier": "FIL",
          "componentUuid": "foo1234",
          "creationDate": "2017-03-01T09:36:01+0100",
          "flows": Array [
            Array [
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
            ],
            Array [
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
            ],
          ],
          "fromHotspot": false,
          "key": "unknown",
          "line": 25,
          "message": "Reduce the number of conditional operators (4) used in the expression",
          "project": "myproject",
          "projectKey": "foo",
          "projectName": "Foo",
          "rule": "javascript:S1067",
          "ruleName": "foo",
          "secondaryLocations": Array [
            Object {
              "component": "main.js",
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
            Object {
              "component": "main.js",
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
          ],
          "severity": "MAJOR",
          "status": "OPEN",
          "textRange": Object {
            "endLine": 26,
            "endOffset": 15,
            "startLine": 25,
            "startOffset": 0,
          },
          "transitions": Array [],
          "type": "BUG",
        }
      }
      issuesByLine={Object {}}
      lastSnippetGroup={false}
      loadDuplications={[Function]}
      locations={Array []}
      onIssueChange={[MockFunction]}
      onIssuePopupToggle={[Function]}
      onIssueSelect={[MockFunction]}
      onLocationSelect={[MockFunction]}
      renderDuplicationPopup={[Function]}
      scroll={[MockFunction]}
      snippetGroup={
        Object {
          "component": Object {},
          "locations": Array [],
          "sources": Array [],
        }
      }
    />
  </ContextProvider>
  <ContextProvider
    key="unknown-0-1"
    value={
      Object {
        "branchLike": undefined,
        "file": Object {
          "canMarkAsFavorite": true,
          "fav": false,
          "key": "project:main.js",
          "longName": "main.js",
          "measures": Object {
            "coverage": "85.2",
            "duplicationDensity": "1.0",
            "issues": "12",
            "lines": "56",
          },
          "name": "main.js",
          "path": "main.js",
          "project": "project",
          "projectName": "MyProject",
          "q": "FIL",
          "uuid": "foo-bar",
        },
      }
    }
  >
    <ComponentSourceSnippetGroupViewer
      duplicationsByLine={Object {}}
      isLastOccurenceOfPrimaryComponent={false}
      issue={
        Object {
          "actions": Array [],
          "component": "test.js",
          "componentLongName": "main.js",
          "componentQualifier": "FIL",
          "componentUuid": "foo1234",
          "creationDate": "2017-03-01T09:36:01+0100",
          "flows": Array [
            Array [
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
            ],
            Array [
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
              Object {
                "component": "main.js",
                "textRange": Object {
                  "endLine": 2,
                  "endOffset": 2,
                  "startLine": 1,
                  "startOffset": 1,
                },
              },
            ],
          ],
          "fromHotspot": false,
          "key": "unknown",
          "line": 25,
          "message": "Reduce the number of conditional operators (4) used in the expression",
          "project": "myproject",
          "projectKey": "foo",
          "projectName": "Foo",
          "rule": "javascript:S1067",
          "ruleName": "foo",
          "secondaryLocations": Array [
            Object {
              "component": "main.js",
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
            Object {
              "component": "main.js",
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
          ],
          "severity": "MAJOR",
          "status": "OPEN",
          "textRange": Object {
            "endLine": 26,
            "endOffset": 15,
            "startLine": 25,
            "startOffset": 0,
          },
          "transitions": Array [],
          "type": "BUG",
        }
      }
      issuesByLine={Object {}}
      lastSnippetGroup={true}
      loadDuplications={[Function]}
      locations={
        Array [
          Object {
            "component": "main.js",
            "index": 0,
            "textRange": Object {
              "endLine": 2,
              "endOffset": 2,
              "startLine": 1,
              "startOffset": 1,
            },
          },
        ]
      }
      onIssueChange={[MockFunction]}
      onIssuePopupToggle={[Function]}
      onIssueSelect={[MockFunction]}
      onLocationSelect={[MockFunction]}
      renderDuplicationPopup={[Function]}
      scroll={[MockFunction]}
      snippetGroup={
        Object {
          "component": Object {
            "canMarkAsFavorite": true,
            "fav": false,
            "key": "project:main.js",
            "longName": "main.js",
            "measures": Object {
              "coverage": "85.2",
              "duplicationDensity": "1.0",
              "issues": "12",
              "lines": "56",
            },
            "name": "main.js",
            "path": "main.js",
            "project": "project",
            "projectName": "MyProject",
            "q": "FIL",
            "uuid": "foo-bar",
          },
          "locations": Array [
            Object {
              "component": "main.js",
              "index": 0,
              "textRange": Object {
                "endLine": 2,
                "endOffset": 2,
                "startLine": 1,
                "startOffset": 1,
              },
            },
          ],
          "sources": Object {
            "16": Object {
              "code": "<span class=\\"k\\">import</span> java.util.<span class=\\"sym-9 sym\\">ArrayList</span>;",
              "coverageStatus": "covered",
              "coveredConditions": 2,
              "duplicated": false,
              "isNew": true,
              "line": 16,
              "scmAuthor": "simon.brandhof@sonarsource.com",
              "scmDate": "2018-12-11T10:48:39+0100",
              "scmRevision": "80f564becc0c0a1c9abaa006eca83a4fd278c3f0",
            },
          },
        }
      }
    />
  </ContextProvider>
</div>
`;
